src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1692,127 +1692,150 @@ private static Guid GenerateGuidFromName(string name)
             {
-                // TODO: how do we want to handle this? For now we ignore it...
-                return null;
+                if (typeCode == TypeCode.Int64)
+                {
+                    return *(long*)dataPointer;
+                }
+                else if (typeCode == TypeCode.UInt64)
+                {
+                    return *(ulong*)dataPointer;
+                }
+                else if (typeCode == TypeCode.Double)
+                {
+                    return *(double*)dataPointer;
+                }
+                else if (typeCode == TypeCode.DateTime)
+                {
+                    return *(DateTime*)dataPointer;
+                }
+                else if (IntPtr.Size == 8 && dataType == typeof(IntPtr))
+                {
+                    return *(IntPtr*)dataPointer;
+                }
             }
             else
             {
-                if (dataType.IsEnum)
+                if (typeCode == TypeCode.Decimal)
                 {
-                    dataType = Enum.GetUnderlyingType(dataType);
-
-                    // Enums less than 4 bytes in size should be treated as int.
-                    switch (Type.GetTypeCode(dataType))
-                    {
-                        case TypeCode.Byte:
-                        case TypeCode.SByte:
-                        case TypeCode.Int16:
-                        case TypeCode.UInt16:
-                            dataType = typeof(int);
-                            break;
-                    }
-                    goto Again;
+                    Debug.Assert(size == 16);
+                    return *(decimal*)dataPointer;
                 }
-
-                // Everything else is marshaled as a string.
-                // ETW strings are NULL-terminated, so marshal everything up to the first
-                // null in the string.
-                if (dataPointer == IntPtr.Zero)
+                else if (dataType == typeof(Guid))
                 {
-                    return null;
+                    Debug.Assert(size == 16);
+                    return *(Guid*)dataPointer;
                 }
+            }
 
-                return new string((char*)dataPointer);
+            if (dataType != typeof(byte*))
+            {
+                goto String;
             }
+
+        BytePtr:
+            byte[] blob = new byte[data->Size];
+            for (int i = 0; i < blob.Length; i++)
+            {
+                blob[i] = *((byte*)dataPointer + i);
+            }
+            return blob;
+
+        String:
+            // Everything else is marshaled as a string.
+            // ETW strings are NULL-terminated, so marshal everything up to the first
+            // null in the string.
+#if DEBUG
+            Debug.Assert(data->Size % 2 == 0, "String size should be even");
+            for (int i = 0; i < data->Size / 2 - 1; i++) Debug.Assert(*((char*)dataPointer + i) != 0, "String may not contain null chars");
```

Comment:
FYI @josalem: this change can be related to https://github.com/dotnet/runtime/issues/52025 for EventPipe.

Reply:
Yup. There's currently nothing stopping users from putting a string with embedded nulls into `EventSource`.